### DISCUSSION TITLE
Kong response is pretty slow if upstream responses body is large.

### DISCUSSION THREAD

**Discussion by samsam2** [Kong/kong#9636] [Help] [answered]
Hi,
We have an upstream container which has REST service and sometimes response with data size around 3.7 MB with approximate time taken is 2 seconds, however, if the same response is routed through kong it takes around 20 mins. Really not sure whats wrong with the configuration, can you please throw some suggestions to improve the response time?

Kong version : 2.6

Below is the route configuration, please note the read timeout is 10 mins we had to change it to 30 mins, that is when we came to know the processing time is taking around 20 mins.
{
            "created_at": 1666086626,
            "updated_at": 1666086626,
            "tags": null,
            "ca_certificates": null,
            "host": "sure-rest",
            "write_timeout": 600000,
            "protocol": "http",
            "name": "sure-rest",
            "tls_verify": null,
            "retries": 3,
            "tls_verify_depth": null,
            "id": "xxxxxxx",
            "port": 80,
            "client_certificate": null,
            "path": "/oss/sure",
            "read_timeout": 600000,
            "connect_timeout": 600000
        }
        
        Thanks.

**Comment by StarlightIbuki** [comment, marked as answer]
Hi. Thank you for reporting this. Could you provide more details about your setup? For example what plugins are in use? Especially those related to responses? And could you reproduce this with 2.8 and 3.0 gateways?
Do you see any unusual logs in error.log? Also, given the really extreme latency, it may help provide access.log(and log from upstream).

**Reply by samsam2** [reply]
Hi,

Given we are seeing issue in customer production server never tired upgrading to 2.8 or 3.0.

Please find some details below:

We use DB-less and declarative configuration.

Plugin details :
We use oauth2-multitenant and token-pep-url plugin, however, the route we are using now has only oauth2-multitenenat.

plugins:
      - config:
          credentials: false
          exposed_headers: null
          headers: null
          max_age: null
          methods: null
          origins:
            - '*'
          preflight_continue: false
        enabled: true
        name: cors
        protocols:
          - http
          - https
      - config:
          allowed_flows: TOKEN,PASSWORD
          application_default_value: null
          application_header: appId
          cookie_path: null
          filters: null
          frontend_address: 135.121.231.104:28443
          gateway_client_id: SURE_APP
          gateway_client_secret: xxxxxxx
          gateway_realm_url: https://wscp-authentication:8666/wavesuite/common/auth/realms/SURE
          logout_path: null
          post_logout_redirect_uri: null
          realm_mapper_base_path: /oss-uam/sure/public
          realm_mapper_host: 127.0.0.1:28443
          ssl_verify: false
          tenant_default_value: null
          tenant_header: tenantId
          token_to_upstream: true
          usergroup_default_value: null
          usergroup_header: ugId
        enabled: true
        name: oauth2-multitenant


kong.conf:

database = off
declarative_config = /opt/wavesuite/etc/kong.yaml

Service and upstream details :
{
            "created_at": 1660051555,
            "host_header": null,
            "tags": null,
            "healthchecks": {
                "active": {
                    "https_sni": null,
                    "http_path": "/",
                    "healthy": {
                        "interval": 0,
                        "http_statuses": [
                            200,
                            302
                        ],
                        "successes": 0
                    },
                    "unhealthy": {
                        "http_statuses": [
                            429,
                            404,
                            500,
                            501,
                            502,
                            503,
                            504,
                            505
                        ],
                        "timeouts": 0,
                        "http_failures": 0,
                        "interval": 0,
                        "tcp_failures": 0
                    },
                    "https_verify_certificate": true,
                    "timeout": 1,
                    "concurrency": 10,
                    "type": "http"
                },
                "passive": {
                    "unhealthy": {
                        "tcp_failures": 0,
                        "http_statuses": [
                            429,
                            500,
                            503
                        ],
                        "timeouts": 0,
                        "http_failures": 0
                    },
                    "healthy": {
                        "successes": 0,
                        "http_statuses": [
                            200,
                            201,
                            202,
                            203,
                            204,
                            205,
                            206,
                            207,
                            208,
                            226,
                            300,
                            301,
                            302,
                            303,
                            304,
                            305,
                            306,
                            307,
                            308
                        ]
                    },
                    "type": "http"
                },
                "threshold": 0
            },
            "hash_on": "none",
            "client_certificate": null,
            "name": "sure-rest",
            "algorithm": "round-robin",
            "id": "XXXXXXX",
            "slots": 10000,
            "hash_on_header": null,
            "hash_on_cookie": null,
            "hash_on_cookie_path": "/",
            "hash_fallback": "none",
            "hash_fallback_header": null
        }

{
            "created_at": 1666086626,
            "updated_at": 1666086626,
            "request_buffering": true,
            "response_buffering": true,
            "regex_priority": 0,
            "hosts": null,
            "https_redirect_status_code": 426,
            "preserve_host": false,
            "name": null,
            "tags": null,
            "paths": [
                "/oss/sure"
            ],
            "methods": null,
            "sources": null,
            "destinations": null,
            "path_handling": "v0",
            "id": "XXXX",
            "headers": null,
            "strip_path": true,
            "service": {
                "id": "XXXXXX"
            },
            "protocols": [
                "http",
                "https"
            ],
            "snis": null
        }

I don't see anything in error log though, request/response sent and received from the  'upsteam' server is pretty fast 2-3 seconds, however, the response to the original client is extremely slow.

access logs:
 info HTTP/1.1
127.0.0.1 31/Oct/2022:08:22:59 +0000 HTTP/1.1 GET /wavesuite/cp/admin/api/v1/server/info 200  GET /wavesuite/cp/admin/api/v1/server/info HTTP/1.1
127.0.0.1 31/Oct/2022:08:23:30 +0000 HTTP/1.1 GET /wavesuite/cp/admin/api/v1/server/info 200  GET /wavesuite/cp/admin/api/v1/server/info HTTP/1.1
127.0.0.1 31/Oct/2022:08:24:00 +0000 HTTP/1.1 GET /wavesuite/cp/admin/api/v1/server/info 200  GET /wavesuite/cp/admin/api/v1/server/info HTTP/1.1
127.0.0.1 31/Oct/2022:08:24:30 +0000 HTTP/1.1 GET /wavesuite/cp/admin/api/v1/server/info 200  GET /wavesuite/cp/admin/api/v1/server/info HTTP/1.1

error logs:

2022/10/28 15:42:22 [warn] 336#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/28 15:42:23 [warn] 349#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/28 16:01:54 [warn] 344#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/28 16:01:55 [warn] 357#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 01:04:38 [warn] 342#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 01:04:38 [warn] 355#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 01:07:36 [warn] 351#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 01:07:36 [warn] 364#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 01:11:40 [warn] 575#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 01:11:40 [warn] 582#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 01:11:40 [notice] 582#0: signal process started
2022/10/29 16:13:26 [warn] 7067#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 16:13:26 [warn] 7074#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 16:13:26 [notice] 7074#0: signal process started
2022/10/29 16:20:29 [warn] 7439#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 16:20:29 [warn] 7446#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 16:20:29 [notice] 7446#0: signal process started
2022/10/29 16:21:06 [warn] 7491#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 16:21:06 [warn] 7498#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 16:21:06 [notice] 7498#0: signal process started
2022/10/29 16:22:41 [warn] 7589#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 16:22:41 [warn] 7596#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/10/29 16:22:41 [notice] 7596#0: signal process started
~                                                           


Please let me know if I can enable any debug logs and get you some more logs.

**Reply by StarlightIbuki** [reply]
Sorry that I don't have much clue after seeing those configs and logs you provided. I suggest trying to find a minimal setup to reproduce the issue (excluding sensitive information, and we could debug outside the production environment). And I still suggest upgrading to 2.8/3.0 to see if the problem persists, as 2.6 is near to EOL.

**Reply by samsam2** [reply]
Hi,
Thanks for the response.
I have shared all the configuration details including whatever logs I am getting, can you please share the details of how to enable debug logs so that I can collect them as you requested?

**Reply by StarlightIbuki** [reply]
You could put this into kong.conf:

```conf
log_level = debug
```

But I don't expect it to help a lot.

**Reply by dndx** [reply]
Could you provide your debug level logs to us?

**Reply by samsam2** [reply]
After further debugging we found there was an issue on upstream server in few scenarios which was causing the problem. We can please close this conversation. Thanks a lot for all the input you have provided.
